### PR TITLE
make sure getindex on DataFrameRows does not alias passed selector

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3191

### CODE REVIEW

**Comment by bkamins** [comment]
This works around the issue that when making a view a copy of `idx` is not made. Unfortunately `copy` is not always doable so it is a bit tricky.

The problem with `filter` in Base Julia is as follows:
```
function filter(f, a::AbstractArray)
    (IndexStyle(a) != IndexLinear()) && return a[map(f, a)::AbstractArray{Bool}]

    j = 1
    idxs = Vector{Int}(undef, length(a))
    for idx in eachindex(a)
        @inbounds idxs[j] = idx
        ai = @inbounds a[idx]
        j = ifelse(f(ai), j+1, j)
    end
    resize!(idxs, j-1)
    res = a[idxs]
    empty!(idxs) # I do not know why this line is needed
    sizehint!(idxs, 0) # I do not know why this line is needed
    return res
end

```

File: src/abstractdataframe/iteration.jl [hunk 1]
```diff
@@ -79,7 +79,7 @@ Base.size(itr::DataFrameRows) = (size(parent(itr), 1), )
 
 Base.@propagate_inbounds Base.getindex(itr::DataFrameRows, i::Int) = parent(itr)[i, :]
 Base.@propagate_inbounds Base.getindex(itr::DataFrameRows, idx) =
-    eachrow(@view parent(itr)[idx, :])
+    eachrow(@view parent(itr)[idx isa AbstractVector{Int} ? copy(idx) : idx, :])
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:82]: Add a comment like this?
```
# `filter(f, ::AbstractArray)` in Base assumes that `idx` can be overwritten after calling `getindex`
```

**Comment by nalimilan** [comment]
Woops. This was introduced by https://github.com/JuliaLang/julia/pull/31929/files#r987877494. Sounds like a good fix at least in the short term. Though it's a bit annoying to have to make all `getindex` calls copy the input index unless that's really required by the `AbstractArray` interface. If not, we could define our own `filter` implementation.

File: src/abstractdataframe/iteration.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:82]: I do not think it is needed. This copying operation should be done anyway. The issue is that `array[idx]` operation should produce a result that does not change after `idx` is changed. So this PR is just a bugfix (that is why I marked it bug).

**Comment by bkamins** [comment]
> Though it's a bit annoying to have to make all `getindex` calls copy the input index

The point is that Base Julia assumes (correctly IMO) that `getindex` produces an independent object from `idx`. We internally, for performance reasons, create a view, so I think it is our responsibility to de-alias (note that for other objects than `AbstractVector{Int}` de-aliasing is ensured by `SubDataFrame` logic).

**Comment by bkamins** [comment]
@nalimilan - even if something is changed in Base Julia, I think we still need this fix for backward versions.

File: src/abstractdataframe/iteration.jl (hunk 1 above)
Comment (nalimilan) [reply src/abstractdataframe/iteration.jl:82]: OK. I'm not sure it's documented but we'll check that later.

BTW, detecting cases where we need to copy is potentially tricky. For example, when indexing with a `PooledArray{UInt}`, conversion to `AbstractArray{Int}` could result in a `PooledArray{Int}` which shares the same reference codes, only with a different pool. Currently we don't do it but that would make sense as an optimization. If we want to be super defensive we could do something like
```suggestion
    eachrow(@view parent(itr)[idx isa AbstractVector && !eltype(idx) <: Bool ? copy(idx) : idx, :])
```

**Review by nalimilan — APPROVED** [review]

**Pull request merged**